lib/idemeumjwt/jwt.go

```diff
@@ -0,0 +1,199 @@
+package idemeumjwt
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/subtle"
+	"fmt"
+	"github.com/form3tech-oss/jwt-go"
+	"github.com/gravitational/trace"
+	"golang.org/x/oauth2"
+	"gopkg.in/square/go-jose.v2"
+	"gopkg.in/square/go-jose.v2/json"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"time"
+)
+
+var TimeFunc = time.Now
+
+type IdemeumClaims struct {
+	Audience   string   `json:"aud,omitempty"`
+	ExpiresAt  int64    `json:"exp,omitempty"`
+	Id         string   `json:"jti,omitempty"`
+	IssuedAt   int64    `json:"iat,omitempty"`
+	Issuer     string   `json:"iss,omitempty"`
+	NotBefore  int64    `json:"nbf,omitempty"`
+	Subject    string   `json:"sub,omitempty"`
+	Roles      []string `json:"roles,omitempty"`
+	SessionTTL int64    `json:"sessionTTL,omitempty"`
+}
+
+func ValidateJwtToken(ServiceToken string, TenantUrl string) (*IdemeumClaims, error) {
+	//Token Validation
+	key, err := LoadPublicKey(TenantUrl + "/.well-known/jwks.json")
+	if err != nil {
+		log.Printf("Failed to load public key for tenant :%v \n", TenantUrl)
+		return nil, trace.BadParameter("invalid configuration")
+	}
+
+	parser := &jwt.Parser{SkipClaimsValidation: true}
+	token, err := parser.ParseWithClaims(ServiceToken, &IdemeumClaims{}, func(token *jwt.Token) (interface{}, error) {
+		return key, nil
+	})
+
+	if err != nil {
+		log.Print("Failed to validate the idemeum token", err)
+		return nil, trace.BadParameter("Invalid idemeum token")
+	}
+
+	claims, ok := token.Claims.(*IdemeumClaims)
+	if !ok && !token.Valid {
+		log.Print("Failed to validate the idemeum token")
+		return nil, trace.BadParameter("invalid token")
+	}
+
+	err = claims.Valid()
+	if err != nil {
+		log.Print("Failed to validate the idemeum token", err)
+		return nil, trace.BadParameter("token invalid")
+	}
+
+	if !claims.VerifyIssuer(TenantUrl, true) {
+		return nil, trace.BadParameter("token invalid Issuer")
+	}
+
+	if !claims.VerifyAudience(TenantUrl, true) {
+		return nil, trace.BadParameter("token invalid audience")
+	}
+	return claims, nil
+}
+
+func doRequest(ctx context.Context, req *http.Request) (*http.Response, error) {
+	client := http.DefaultClient
+	if c, ok := ctx.Value(oauth2.HTTPClient).(*http.Client); ok {
+		client = c
+	}
+	return client.Do(req.WithContext(ctx))
+}
+
+func LoadPublicKey(jwksURL string) (*ecdsa.PublicKey, error) {
```

Comment:
All package methods start with lower chase character. First uppercase letter makes it public methods.

Make package scope